src/Honeycomb/Api/Events.hs

```diff
@@ -11,29 +11,62 @@ import qualified Data.ByteString.Lazy as LBS
 import Data.Coerce (coerce)
 import qualified Data.Text as T
 import qualified Data.Text.Encoding as TE
+import GHC.Int (Int64)
 import Honeycomb.Api.Types
 import Lens.Micro ((^.))
 import qualified Network.HTTP.Client as Client
 import Network.HTTP.Types.Header (RequestHeaders)
 import Network.URI (normalizeEscape)
 
+maxEntryLength :: Int64
+maxEntryLength = 100000
+
+maxBodyLength :: Int64
+maxBodyLength = 5000000
+
 sendEvents ::
   MonadIO m =>
   (Client.Request -> m (Client.Response LBS.ByteString)) ->
   RequestOptions ->
   [Event] ->
-  m (Client.Response (Maybe [Integer]))
+  m SendEventsResponse
 sendEvents httpLbs requestOptions events = do
   initReq <- liftIO $ Client.parseRequest batchUri
-  response <- httpLbs $ newReq initReq
-  pure $ JSON.decode' <$> response
+  let (unsent, dropped, body) = createBodyFromEvents True events [] "[" 1
+  response <- httpLbs $ newReq initReq $ body <> "]"
+  pure SendEventsResponse
+    { unsentEvents = unsent,
+      oversizedEvents = dropped,
+      serviceResponse = JSON.decode <$> response
+    }
   where
-    newReq :: Client.Request -> Client.Request
-    newReq initReq =
+    createBodyFromEvents :: Bool -> [Event] -> [Event] -> LBS.ByteString -> Int64 -> ([Event], [Event], LBS.ByteString)
+    createBodyFromEvents isFirst toSend dropped bs bsLength =
+      case toSend of
+        [] -> ([], dropped, bs)
+        l@(hd : tl) ->
+          let newEncoded = JSON.encode hd
+              newEncodedLength = LBS.length newEncoded
+              newBS =
+                if isFirst
+                  then bs <> JSON.encode hd
+                  else bs <> "," <> JSON.encode hd
```

Comment:
Have you looked at https://hackage.haskell.org/package/bytestring-0.10.10.0/docs/Data-ByteString-Builder.html
IIRC Builders are preferred for incremental construction and Aeson might even have a direct interface into Builder encoding

Reply:
Ah I didn't; I'll update to that! Sounds worth it.